rmc/utils/generic.py

```diff
@@ -220,6 +220,122 @@ def process_context_ht(
+        protein_start=ht.transcript_consequences.protein_start,
+        protein_end=ht.transcript_consequences.protein_end,
+        amino_acids=ht.transcript_consequences.amino_acids,
+    )
+
+    if keep_transcripts:
+        logger.info("Filtering to desired transcripts only...")
+        ht = ht.filter(hl.literal(keep_transcripts).contains(ht.transcript))
+
+    ht = ht.naive_coalesce(n_partitions)
+    ht = ht.checkpoint(
+        f"{TEMP_PATH_WITH_FAST_DEL}/rmc/amino_acids.ht",
+        _read_if_exists=not overwrite_temp,
+        overwrite=overwrite_temp,
+    )
+    return ht
+
+
+def get_ref_aa(
+    ht: hl.Table,
+    overwrite_temp: bool,
+    extra_aa_map: Dict[str, str] = {"*": "Ter", "U": "Sec"},
+    aa_to_remove: str = "X",
+) -> hl.Table:
+    """
+    Filter input HT to keep only reference amino acid information.
+
+    :param ht: Input Table. Should be HT output from `get_aa_from_context`.
+    :param overwrite_temp: Whether to overwrite temporary data.
+        If False, will read existing temp data rather than overwriting.
+        If True, will overwrite temp data.
+    :param extra_aa_map: Dictionary mapping any amino acid one letter codes to three letter codes.
+        Designed to capture any amino acids not present in `ACID_NAMES_PATH`.
+        Default is {"*": "Ter", "U": "Sec"}.
+    :param aa_to_remove: Any amino acid one letter codes to remove. Default is "X".
+    :return: HT with one reference amino acid annotated per locus/transcript if `collect_all_aa` is False.
+        HT grouped by transcript with a list of all reference amino acids in that transcript if `collect_all_aa` is True.
+    """
+    logger.info("Mapping amino acid one letter codes to three letter codes...")
+    aa_map = get_aa_map()
+    # Add any extra mappings into amino acid mapping dict
+    if extra_aa_map:
+        aa_map = aa_map.update(extra_aa_map)
+    ht = ht.annotate(ref_aa_1_letter=ht.amino_acids.split("/")[0])
+    ht = ht.annotate(
+        ref_aa=hl.literal(aa_map).get(ht.ref_aa_1_letter, ht.ref_aa_1_letter)
+    )
+    if aa_to_remove:
+        ht = ht.filter(ht.ref_aa != aa_to_remove)
+
+    # Double check that protein start always equals protein end
+    protein_num_check = ht.aggregate(
+        hl.agg.count_where(ht.protein_start != ht.protein_end)
+    )
+    if protein_num_check != 0:
+        raise DataException(
+            f"{protein_num_check} sites had different protein start and end values --"
+            " please double check!"
+        )
+    # Reformat reference AA to have both the 3 letter code and number
+    ht = ht.annotate(ref_aa=hl.format("%s%s", ht.ref_aa, ht.protein_start))
+
+    # Select fields and checkpoint
+    ht = ht.select("ref_aa", "transcript")
+    ht = ht.checkpoint(
+        f"{TEMP_PATH_WITH_FAST_DEL}/rmc/ref_amino_acids.ht",
+        _read_if_exists=not overwrite_temp,
+        overwrite=overwrite_temp,
+    )
+
+    # Collect by key to collapse AA per locus
```

Comment:
> It would shrink the table too.

that's a good point! I didn't add it to the top of the function so that I could (lazily) work with unnested annotations


> Also have you checked that when AA info is missing, it is missing for all rows for a locus-transcript combo? I.e. not just missing for one variant which we happen to pick up in this collapse. 

ah, I have not checked this

Reply:
I might keep the collect by key where it is to make the ref aa check/protein number check more straightforward

Reply:
Ok, if you're not worried about the cost!